File: src/api/modules/library.ts
import { invoke } from '@tauri-apps/api/core'
import type { ApiResponse } from './types'
import type { Song } from '../../types'
import { getSongs } from './song'

// 喜欢
export async function getLikedPaths(): Promise<string[]> {
  const response = await invoke<ApiResponse<string[]>>('get_liked_paths')
  if (!response.success) throw new Error(response.error)
  return response.data || []
}

export async function getLikedSongs(): Promise<{ songs: Song[] }> {
  const [allSongs, likedPaths] = await Promise.all([
    getSongs(),
    getLikedPaths(),
  ])
  const likedSongs = allSongs.filter(song => likedPaths.includes(song.path))
  return { songs: likedSongs }
}

export async function toggleLike(path: string, liked: boolean): Promise<void> {
  const response = await invoke<ApiResponse<void>>('toggle_like', { path, liked })
  if (!response.success) throw new Error(response.error)
}

export async function isSongLiked(path: string): Promise<boolean> {
  const response = await invoke<ApiResponse<boolean>>('is_song_liked', { path })
  if (!response.success) throw new Error(response.error)
  return response.data || false
}

// 隐藏
export async function hideSong(path: string, isAuto?: boolean): Promise<void> {
  const response = await invoke<ApiResponse<void>>('hide_song', { path, isAuto })
  if (!response.success) throw new Error(response.error)
}

export async function unhideSong(path: string): Promise<void> {
  const response = await invoke<ApiResponse<void>>('unhide_song', { path })
  if (!response.success) throw new Error(response.error)
}

export async function getHiddenPaths(): Promise<string[]> {
  const response = await invoke<ApiResponse<string[]>>('get_hidden_paths')
  if (!response.success) throw new Error(response.error)
  return response.data || []
}

export async function getHiddenSongs(): Promise<Song[]> {
  const response = await invoke<ApiResponse<Song[]>>('get_hidden_songs')
  if (!response.success) throw new Error(response.error)
  return response.data || []
}

export async function hideSongsBatch(paths: string[], isAuto?: boolean): Promise<number> {
  const response = await invoke<ApiResponse<number>>('hide_songs_batch', { paths, isAuto })
  if (!response.success) throw new Error(response.error)
  return response.data || 0
}

export async function unhideSongsBatch(paths: string[]): Promise<number> {
  const response = await invoke<ApiResponse<number>>('unhide_songs_batch', { paths })
  if (!response.success) throw new Error(response.error)
  return response.data || 0
}

export async function clearHiddenSongs(): Promise<number> {
  const response = await invoke<ApiResponse<number>>('clear_hidden_songs')
  if (!response.success) throw new Error(response.error)
  return response.data || 0
}

export async function getHiddenCount(): Promise<number> {
  const response = await invoke<ApiResponse<number>>('get_hidden_count')
  if (!response.success) throw new Error(response.error)
  return response.data || 0
}

export async function isSongHidden(path: string): Promise<boolean> {
  const response = await invoke<ApiResponse<boolean>>('is_song_hidden', { path })
  if (!response.success) throw new Error(response.error)
  return response.data || false
}

File: src-tauri/src/commands/library.rs
use tauri::State;

use crate::database::Database;
use crate::database::Song;
use super::common::{ApiResponse, validate_path_in_music_folder, MAX_BATCH_SIZE};

#[tauri::command]
pub async fn hide_song(
    db: State<'_, Database>,
    path: String,
    is_auto: Option<bool>,
) -> Result<ApiResponse<()>, String> {
    if let Err(e) = validate_path_in_music_folder(&db, &path).await {
        return Ok(ApiResponse::err(e));
    }

    match db.hide_song(&path, is_auto.unwrap_or(false)).await {
        Ok(_) => Ok(ApiResponse::ok(())),
        Err(e) => Ok(ApiResponse::err(e)),
    }
}

#[tauri::command]
pub async fn unhide_song(
    db: State<'_, Database>,
    path: String,
) -> Result<ApiResponse<()>, String> {
    if let Err(e) = validate_path_in_music_folder(&db, &path).await {
        return Ok(ApiResponse::err(e));
    }

    match db.unhide_song(&path).await {
        Ok(_) => Ok(ApiResponse::ok(())),
        Err(e) => Ok(ApiResponse::err(e)),
    }
}

#[tauri::command]
pub async fn get_hidden_paths(db: State<'_, Database>) -> Result<ApiResponse<Vec<String>>, String> {
    match db.get_hidden_paths().await {
        Ok(paths) => Ok(ApiResponse::ok(paths)),
        Err(e) => Ok(ApiResponse::err(e)),
    }
}

#[tauri::command]
pub async fn get_hidden_songs(db: State<'_, Database>) -> Result<ApiResponse<Vec<Song>>, String> {
    match db.get_hidden_songs().await {
        Ok(songs) => Ok(ApiResponse::ok(songs)),
        Err(e) => Ok(ApiResponse::err(e)),
    }
}

#[tauri::command]
pub async fn hide_songs_batch(
    db: State<'_, Database>,
    paths: Vec<String>,
    is_auto: Option<bool>,
) -> Result<ApiResponse<usize>, String> {
    if paths.len() > MAX_BATCH_SIZE {
        return Ok(ApiResponse::err(format!(
            "Too many paths (max {})", MAX_BATCH_SIZE
        )));
    }

    match db.hide_songs_batch(paths, is_auto.unwrap_or(false)).await {
        Ok(count) => Ok(ApiResponse::ok(count)),
        Err(e) => Ok(ApiResponse::err(e)),
    }
}

#[tauri::command]
pub async fn unhide_songs_batch(
    db: State<'_, Database>,
    paths: Vec<String>,
) -> Result<ApiResponse<usize>, String> {
    if paths.len() > MAX_BATCH_SIZE {
        return Ok(ApiResponse::err(format!(
            "Too many paths (max {})", MAX_BATCH_SIZE
        )));
    }

    match db.unhide_songs_batch(paths).await {
        Ok(count) => Ok(ApiResponse::ok(count)),
        Err(e) => Ok(ApiResponse::err(e)),
    }
}

#[tauri::command]
pub async fn clear_hidden_songs(db: State<'_, Database>) -> Result<ApiResponse<usize>, String> {
    match db.clear_hidden_songs().await {
        Ok(count) => Ok(ApiResponse::ok(count)),
        Err(e) => Ok(ApiResponse::err(e)),
    }
}

#[tauri::command]
pub async fn get_hidden_count(db: State<'_, Database>) -> Result<ApiResponse<i64>, String> {
    match db.get_hidden_count().await {
        Ok(count) => Ok(ApiResponse::ok(count)),
        Err(e) => Ok(ApiResponse::err(e)),
    }
}

#[tauri::command]
pub async fn is_song_hidden(
    db: State<'_, Database>,
    path: String,
) -> Result<ApiResponse<bool>, String> {
    if let Err(e) = validate_path_in_music_folder(&db, &path).await {
        return Ok(ApiResponse::err(e));
    }

    match db.is_song_hidden(&path).await {
        Ok(hidden) => Ok(ApiResponse::ok(hidden)),
        Err(e) => Ok(ApiResponse::err(e)),
    }
}
